File: backends/springboot/src/main/java/com/example/todobench/support/TodoPagination.java
package com.example.todobench.support;

/**
 * Keyset pagination options for GET /todos, aligned with the Express reference backend.
 *
 * - No limit/afterId query params: unbounded list (legacy compatibility).
 * - Any limit or afterId param present: paginated mode using WHERE id > afterId.
 */
public record TodoPagination(boolean paginated, Integer limit, int afterId) {

    public static final int MAX_LIMIT = 500;
    public static final int DEFAULT_LIMIT = 100;

    public static TodoPagination fromQueryParams(String limitParam, String afterIdParam) {
        boolean hasLimit = limitParam != null;
        boolean hasAfterId = afterIdParam != null;

        if (!hasLimit && !hasAfterId) {
            return new TodoPagination(false, null, 0);
        }

        int requestedLimit = parsePositiveInteger(limitParam, DEFAULT_LIMIT);
        int limit = Math.min(requestedLimit, MAX_LIMIT);
        int afterId = parseNonNegativeInteger(afterIdParam, 0);

        return new TodoPagination(true, limit, afterId);
    }

    private static int parsePositiveInteger(String rawValue, int fallback) {
        if (rawValue == null || rawValue.isBlank()) {
            return fallback;
        }

        try {
            int parsed = Integer.parseInt(rawValue.trim());
            if (parsed <= 0) {
                return fallback;
            }
            return parsed;
        } catch (NumberFormatException error) {
            return fallback;
        }
    }

    private static int parseNonNegativeInteger(String rawValue, int fallback) {
        if (rawValue == null || rawValue.isBlank()) {
            return fallback;
        }

        try {
            int parsed = Integer.parseInt(rawValue.trim());
            if (parsed < 0) {
                return fallback;
            }
            return parsed;
        } catch (NumberFormatException error) {
            return fallback;
        }
    }
}

File: backends/aspnet/Pagination/TodoPagination.cs
namespace TodoBench.AspNet.Pagination;

/// <summary>
/// Keyset pagination options for GET /todos, aligned with the Express reference backend.
/// </summary>
public sealed record TodoPagination(bool Paginated, int? Limit, int AfterId)
{
    public const int MaxLimit = 500;
    public const int DefaultLimit = 100;

    public static TodoPagination FromQueryParams(string? limitParam, string? afterIdParam)
    {
        var hasLimit = limitParam is not null;
        var hasAfterId = afterIdParam is not null;

        if (!hasLimit && !hasAfterId)
        {
            return new TodoPagination(Paginated: false, Limit: null, AfterId: 0);
        }

        var requestedLimit = ParsePositiveInteger(limitParam, DefaultLimit);
        var limit = Math.Min(requestedLimit, MaxLimit);
        var afterId = ParseNonNegativeInteger(afterIdParam, 0);

        return new TodoPagination(Paginated: true, Limit: limit, AfterId: afterId);
    }

    private static int ParsePositiveInteger(string? rawValue, int fallback)
    {
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return fallback;
        }

        if (!int.TryParse(rawValue, out var parsed) || parsed <= 0)
        {
            return fallback;
        }

        return parsed;
    }

    private static int ParseNonNegativeInteger(string? rawValue, int fallback)
    {
        if (string.IsNullOrWhiteSpace(rawValue))
        {
            return fallback;
        }

        if (!int.TryParse(rawValue, out var parsed) || parsed < 0)
        {
            return fallback;
        }

        return parsed;
    }
}
